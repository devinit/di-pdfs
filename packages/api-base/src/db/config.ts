import {TConfig} from 'pg-promise';

// get the connection details from Allan or Donata

export const dwConfig: TConfig = {
    user: process.env.DB_USER,
    database: process.env.DB_DATABASE,
    host: process.env.DB_HOST,
    port: Number(process.env.DB_PORT),
    password: process.env.DB_PASSWORD
};
