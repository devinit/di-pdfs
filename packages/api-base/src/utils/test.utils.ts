// a patch for pretty format output that will exclude uids from storyshots
// this is because they are not constants and change on every new storyshots
// the issue with this is that we cant compare new and old storyshots
import * as prettyFormat from 'pretty-format';

export type ListWithUid = Array<{uid: string} & any>;

export interface IHasUid {
    [field: string]: {
        data: ListWithUid;
        toolTip: any;
    };
}

export const replaceUidInList = (data: ListWithUid): ListWithUid => {
    return data.map(obj => {
        if (obj && obj.uid) obj.uid = 'unique_id_stub';
        return obj;
    });
};

export const uidPatchForTabData = (data: IHasUid) => {
    return Object.keys(data).reduce((acc, key: string) => {
        const itemsWithUid: ListWithUid = data[key].data;
        const newItems =  replaceUidInList(itemsWithUid);
        return {...acc, [key]: { ...data[key], data: newItems}};
    }, {});
};

export const prettyLists =
    (data: ListWithUid) => prettyFormat(replaceUidInList(data));
