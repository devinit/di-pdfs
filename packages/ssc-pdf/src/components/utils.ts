import dvptColorMap from '@devinit/ssc-api/lib/modules/shared/config';

export const getLegendData = (data: object): Array<{color: string; value: string | number}> =>
    Object.keys(data)
            .map(value => ({color: data[value], value}));

export const dvptCoLegendData = (country: string): Array<{color: string; value: string}> => {
    return Object.keys(dvptColorMap[country]).map(value => {
        const color = dvptColorMap[country][value];
        return {color, value};
    });
};
