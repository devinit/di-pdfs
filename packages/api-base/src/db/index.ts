import {IMain, IDatabase, IOptions} from 'pg-promise';
import {dwConfig} from './config';
import * as pgPromise from 'pg-promise';
import * as LRU from 'lru-cache';

export interface IExtensions {
    manyCacheable: (query: string, values?: object) => Promise<any[]>;
}

const lruOpts: LRU.Options<any> = {
    max: 300,
    maxAge: 1000 * 60 * 60 * 60
};

export const dbCache: LRU.Cache<any, any> = LRU(lruOpts);

// pg-promise initialization options:
const options: IOptions<IExtensions> = {
    // Extending the database protocol with our custom modules
    // API: http://vitaly-t.github.io/pg-promise/global.html#event:extend
    extend: (obj: IExtensions & IDatabase<IExtensions>) => {
        obj.manyCacheable = (query: string, values?: any) => {
            const getQuery = values ? pgPromise.as.format(query, values) : query;
            if (dbCache.has(getQuery)) {
                return Promise.resolve(dbCache.get(getQuery));
            }
            return obj.any(getQuery);
        };
    },
    // caching
    receive: (data, _result, event) => {
        // cache recieved
        if (!dbCache.has(event.query)) dbCache.set(event.query, data);
    }
};

const pgp: IMain = pgPromise(options);

// Create the database instance with extensions:
const db = pgp<IExtensions>(dwConfig);

process.on('exit', (code) => {
  // kill db
  pgp.end();
  console.log(`About to exit with code: ${code}, closed DB connection`);
  process.kill(process.pid);
});

process.on('SIGINT', () => {
    // kill db
    pgp.end();
    console.log('Ctrl-C... forced termination closed DB connection');
    process.kill(process.pid);
});

export type IDB = IDatabase<IExtensions> & IExtensions; // necessary repition to make typescript happy
// If you ever need access to the library's root (pgp object), you can do it via db.$config.pgp
// See: http://vitaly-t.github.io/pg-promise/Database.html#.$config
export default db;
