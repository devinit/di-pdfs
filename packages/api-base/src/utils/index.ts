import * as R from 'ramda';
import {IEntity, ICurrency, getCurrency, getEntityBySlugAsync} from '../cms/modules/global';
import * as shortid from 'shortid';
import {IhasDiId, Isummable} from './types';
import {IDatabase} from 'pg-promise';
import {IExtensions} from '../db';

export type IDB = IDatabase<IExtensions> & IExtensions;

export const RECIPIENT = 'recipient';
export const DONOR = 'donor';
export const MULTILATERAL = 'multilateral';
export const  CROSSOVER = 'crossover';
// TODO: should get from a refrence file
export const  budgetTypesRefs = {
    actual: 'Actual',
    estimated: 'Estimated',
    proposed: 'Proposed',
    approved: 'Approved',
    budget: 'Budget',
    proj: 'Projected'
};

// makes numerical like values numbers in an object
export const toNumericFields: (obj: any) => any = (obj) => {
    return R.keys(obj).reduce((newObj: any, key: string) => {
        const isKeyNumerical = Number(obj[key]) ? true : false;
        if (isKeyNumerical) {
            return {...newObj, [key]:  Number(obj[key])};
        }
        return {...newObj, [key]: obj[key]};
    }, {});
};

export const toId: (obj: IhasDiId ) => any = (obj) => {
    if (!obj.di_id) return obj;
    const id = obj.di_id;
    const newObj = R.omit(['di_id'], obj);
    return {...newObj, id };
};

export const getTotal = (data: Isummable[]): number =>
    R.reduce<Isummable, number>((sum: number, obj: Isummable): number => {
        if (obj.value) sum += Number(obj.value);
        return sum;
    }, 0, data);

export const valuesIntoPercents = <T extends {value: number}>(data: T[], precision?: number): T[] => {
    const sum = getTotal(data);
    return data.map((obj: T) => {
        const newValue  = (obj.value / sum) * 100;
        const value = precision ? Number(formatNumbers(newValue, precision)) : Math.round(newValue);
        return Object.assign(obj, {value});
    });
};

export const groupedValuesIntoPercents =
    <T extends {value: number}>(data: T[], groupName: string, precision?: number): T[] => {
    const groups = R.groupBy(R.prop(groupName), data);
    const groupAsPers: any[][] = R.keys(groups).map(key =>
        valuesIntoPercents(groups[key], precision));
    return R.flatten(groupAsPers);
};

export const getCurrencyCode = async (id: string): Promise<string>  => {
    try {
        const currencyList: ICurrency[] = await getCurrency();
        const entity: IEntity | undefined = await getEntityBySlugAsync(id);
        if (!entity) throw new Error(`entity was not found for slug: ${id}`);
        const currency: ICurrency | undefined = R.find(R.propEq('id', entity.id), currencyList) as ICurrency;
        return currency ? currency.code : 'NCU';
    } catch (error) {
        throw error;
    }
};

export const getTableNameFromSql = (sqlStr: string): string | Error => {
    const matches = sqlStr.match(/FROM(.*)WHERE/);
    if (matches && matches.length) {
        return matches[0].split(/\s/)[1];
    }
    return new Error(`couldnt get table name from sql string ${sqlStr}`);
};

export const getIndicatorDataSimple = async<T extends {}> ({db, query}: {db: IDB, query: string}): Promise<T[]> => {
    const raw = await db.manyCacheable(query);
    return raw
        .map(toNumericFields)
        .map(obj => ({...obj, uid: shortid.generate()})) as T[];
};

export const isDonor = async (slug: string): Promise<boolean>  => {
    const obj: IEntity | undefined = await getEntityBySlugAsync(slug);
    if (!obj) throw new Error('Error in isDonor function, entity is undefined');
    if (obj.donor_recipient_type === DONOR) return true;
    return false;
};

export const normalizeKeyName = (columnName: string): string => {
    const str = columnName.includes('value_') ? columnName.split(/value\_/)[1] : columnName;
    return str.replace(/\_/g, '-');
};

export const normalizeKeyNames = (obj: {}) => {
    return R.keys(obj).reduce((acc, key) => {
        const newKeyName = key.includes('_') ? normalizeKeyName(key) : key;
        if (newKeyName) {
            const newObj = R.omit([key], obj);
            return {...newObj, [newKeyName]: obj[newKeyName]};
        }
        return {...acc, [key]: obj[key]}; // return to default
    }, {});
};

export const getCurrentYear = (): number => {
    const date = new Date();
    return date.getFullYear();
};

export const capitalize = (val: string) =>
    `${val[0].toUpperCase()}${R.drop(1, val)}`;

export const addSuffix = (val) => {
    // Borrowed from old datahub codebase
    const lastDigit = (val % 10);
    const lastTwoDigits = (val % 100);
    const suffixObj = ['th', 'st', 'nd', 'rd', 'th', 'th', 'th', 'th', 'th', 'th'];
    const exceptions = [11, 12, 13];
    if (exceptions.indexOf(lastTwoDigits) === -1) {
        return val + suffixObj[lastDigit];
    }
    return val + 'th';
};

export const getMaxAndMin = (data: Array<{year?: number | null}>): number[] => {
    const years = data
        .map(obj => {
            if (obj && obj.year) return Number(obj.year);
            return null;
        })
        .filter(year => year !== null);
    if (!years) return [0, 0];
    const max: number = Math.max.apply(null, years);
    const min: number = Math.min.apply(null, years);
    return [max, min];
};

const removeTrailingZero = (value: string): string => {
    const val = Number(value);
    return  Math.round(val) === val ? val.toString() : value;
};

// (10 ** length) == Math.pow(10, length);
export const roundNum = (num, length): string =>
    (Math.round(num * (10 ** length)) / (10 ** length)).toFixed(length);

export const formatNumbers =
    (value: number | string | undefined | null,
     precision: number = 1,
     shouldrRemoveTrailingZero: boolean = false): string => {
    if (value === undefined || value === null) return 'No data';
    const val = Number(value);
    const absValue = Math.abs(val);
    if (absValue < 1e3) {
        const fixed = roundNum(val, precision);
        return shouldrRemoveTrailingZero ? `${removeTrailingZero(fixed)}` : fixed;
    } else if (absValue >= 1e3 && absValue < 1e6) {
        const newValue = val / 1e3;
        const fixed = roundNum(newValue, precision);
        return shouldrRemoveTrailingZero ? `${removeTrailingZero(fixed)}k` : `${fixed}k`;
    } else if (absValue >= 1e6 && absValue < 1e9) {
        const newValue = val / 1e6;
        const fixed = roundNum(newValue, precision);
        return shouldrRemoveTrailingZero ? `${removeTrailingZero(fixed)}m` : `${fixed}m`;
    } else {
        const newValue = val / 1e9;
        const fixed = roundNum(newValue, precision);
        return shouldrRemoveTrailingZero ? `${removeTrailingZero(fixed)}bn` : `${fixed}bn`;
    }
};
