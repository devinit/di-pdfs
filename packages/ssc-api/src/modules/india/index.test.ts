
import {India} from '.';
import {prettyLists} from '@di-pdfs/api-base/lib/utils/test.utils';
import db from '@di-pdfs/api-base/lib/db';

describe('India', () => {
    const india = new India(db);
    it('getting dvptCooperation', async () => {
       const data = await india.dvptCooperation();
       expect(prettyLists(data)).toMatchSnapshot();
    }, 100000);
    it('getting multiCooperation', async () => {
        const data = await india.multiCooperation();
        expect(prettyLists(data)).toMatchSnapshot();
     }, 100000);
    it('getting tch top recipients', async () => {
        const data = await india.topTchRecipients();
        expect(prettyLists(data)).toMatchSnapshot();
     }, 100000);
    afterAll(() => {
       db.$config.pgp.end();
    });
});
