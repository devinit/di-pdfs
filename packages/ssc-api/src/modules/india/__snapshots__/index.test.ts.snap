// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`India getting dvptCooperation 1`] = `
"Array [
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Interest subsidy for lines of credit\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 39.2,
    \\"year\\": 2008,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 5.1,
    \\"year\\": 2008,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"International organisations\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 233.5,
    \\"year\\": 2008,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"id\\": \\"Loans and advances to foreign governments\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 160.4,
    \\"year\\": 2008,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"id\\": \\"Technical & economic cooperation with other countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 375.7,
    \\"year\\": 2008,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Interest subsidy for lines of credit\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 29.6,
    \\"year\\": 2009,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 4,
    \\"year\\": 2009,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"International organisations\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 137.1,
    \\"year\\": 2009,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"id\\": \\"Loans and advances to foreign governments\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 119.1,
    \\"year\\": 2009,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"id\\": \\"Technical & economic cooperation with other countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 419.7,
    \\"year\\": 2009,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Interest subsidy for lines of credit\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 23.5,
    \\"year\\": 2010,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 4.5,
    \\"year\\": 2010,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"International organisations\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 116.6,
    \\"year\\": 2010,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"id\\": \\"Loans and advances to foreign governments\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 87.8,
    \\"year\\": 2010,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"id\\": \\"Technical & economic cooperation with other countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 451.9,
    \\"year\\": 2010,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Interest subsidy for lines of credit\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 23.5,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 6.7,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"International organisations\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 145.2,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"id\\": \\"Loans and advances to foreign governments\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 153,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"id\\": \\"Technical & economic cooperation with other countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 424,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Interest subsidy for lines of credit\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 40.4,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 7.2,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"International organisations\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 134.2,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"id\\": \\"Loans and advances to foreign governments\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 262.8,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"id\\": \\"Technical & economic cooperation with other countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 538.2,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Interest subsidy for lines of credit\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 60.7,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 8.7,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"International organisations\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 157.9,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"id\\": \\"Loans and advances to foreign governments\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 270.1,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"id\\": \\"Technical & economic cooperation with other countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 776.7,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Interest subsidy for lines of credit\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 60.5,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 8.4,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"International organisations\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 165.9,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"id\\": \\"Loans and advances to foreign governments\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 297.7,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"id\\": \\"Technical & economic cooperation with other countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 717.6,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"id\\": \\"Interest subsidy for lines of credit\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 73.3,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"id\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 3.6,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"id\\": \\"International organisations\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 256.1,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"id\\": \\"Loans and advances to foreign governments\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 445.7,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"id\\": \\"Technical & economic cooperation with other countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 733.2,
    \\"year\\": 2015,
  },
]"
`;

exports[`India getting multiCooperation 1`] = `
"Array [
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"multilateral\\": \\"Asian Development Bank\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 0.7,
    \\"year\\": 2008,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"multilateral\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 17.7,
    \\"year\\": 2008,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"multilateral\\": \\"Southern\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 8.1,
    \\"year\\": 2008,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"multilateral\\": \\"UN\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 206.4,
    \\"year\\": 2008,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"multilateral\\": \\"World Bank Group\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 0.7,
    \\"year\\": 2008,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"multilateral\\": \\"Asian Development Bank\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 0.4,
    \\"year\\": 2009,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"multilateral\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 19.9,
    \\"year\\": 2009,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"multilateral\\": \\"Southern\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 8.7,
    \\"year\\": 2009,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"multilateral\\": \\"UN\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 107.7,
    \\"year\\": 2009,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"multilateral\\": \\"World Bank Group\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 0.4,
    \\"year\\": 2009,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"multilateral\\": \\"Asian Development Bank\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 28,
    \\"year\\": 2010,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"multilateral\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 16.2,
    \\"year\\": 2010,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"multilateral\\": \\"Southern\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 12.7,
    \\"year\\": 2010,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"multilateral\\": \\"UN\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 58.4,
    \\"year\\": 2010,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"multilateral\\": \\"World Bank Group\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 1.3,
    \\"year\\": 2010,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"multilateral\\": \\"Asian Development Bank\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 35.1,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"multilateral\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 14.4,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"multilateral\\": \\"Southern\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 11.1,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"multilateral\\": \\"UN\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 57,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"multilateral\\": \\"World Bank Group\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 27.7,
    \\"year\\": 2011,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"multilateral\\": \\"Asian Development Bank\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 20.4,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"multilateral\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 15.9,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"multilateral\\": \\"Southern\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 7.9,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"multilateral\\": \\"UN\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 56.2,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"multilateral\\": \\"World Bank Group\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 33.8,
    \\"year\\": 2012,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"multilateral\\": \\"Asian Development Bank\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 18.5,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"multilateral\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 21.4,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"multilateral\\": \\"Southern\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 6.3,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"multilateral\\": \\"UN\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 59.4,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"multilateral\\": \\"World Bank Group\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 52.3,
    \\"year\\": 2013,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"multilateral\\": \\"Asian Development Bank\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 33.9,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"multilateral\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 22,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"multilateral\\": \\"Southern\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 13.7,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"multilateral\\": \\"UN\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 56.4,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"multilateral\\": \\"World Bank Group\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 39.9,
    \\"year\\": 2014,
  },
  Object {
    \\"color\\": \\"#bc2629\\",
    \\"multilateral\\": \\"Asian Development Bank\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 10.2,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#8f1b13\\",
    \\"multilateral\\": \\"Other\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 25.5,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#f39b95\\",
    \\"multilateral\\": \\"Southern\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 122,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#e8443a\\",
    \\"multilateral\\": \\"UN\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 72.9,
    \\"year\\": 2015,
  },
  Object {
    \\"color\\": \\"#fbd7cb\\",
    \\"multilateral\\": \\"World Bank Group\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 25.5,
    \\"year\\": 2015,
  },
]"
`;

exports[`India getting tch top recipients 1`] = `
"Array [
  Object {
    \\"recipient\\": \\"Bhutan\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 773,
  },
  Object {
    \\"recipient\\": \\"Afghanistan\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 126.9,
  },
  Object {
    \\"recipient\\": \\"Sri Lanka\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 64.7,
  },
  Object {
    \\"recipient\\": \\"Nepal\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 46.6,
  },
  Object {
    \\"recipient\\": \\"African Countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 37.4,
  },
  Object {
    \\"recipient\\": \\"Bangladesh\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 25.2,
  },
  Object {
    \\"recipient\\": \\"Myanmar\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 17.2,
  },
  Object {
    \\"recipient\\": \\"Other Developing Countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 13.7,
  },
  Object {
    \\"recipient\\": \\"Maldives\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 7.2,
  },
  Object {
    \\"recipient\\": \\"Eurasian Countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 2.6,
  },
  Object {
    \\"recipient\\": \\"Latin American Countries\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 2.2,
  },
  Object {
    \\"recipient\\": \\"Mongolia\\",
    \\"uid\\": \\"unique_id_stub\\",
    \\"value\\": 0.9,
  },
]"
`;
